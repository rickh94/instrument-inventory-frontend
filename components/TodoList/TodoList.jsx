import React, { useState, useEffect, useContext } from 'react'
import PropTypes from 'prop-types'
import {
  Table,
  TableRow,
  TableHead,
  TableCell,
  TableBody,
  TextField,
  Button,
} from '@material-ui/core'
import CheckCircleOutline from '@material-ui/icons/CheckCircleOutline'
import CheckCircle from '@material-ui/icons/CheckCircle'
import AddIcon from '@material-ui/icons/Add'
import DeleteIcon from '@material-ui/icons/Delete'
import ArrowForwardIcon from '@material-ui/icons/ArrowForward'
import EditIcon from '@material-ui/icons/Edit'

import { LoadingHeader } from '..'
import { API } from 'aws-amplify'
import TooltipIconButton from '../TooltipIconButton/TooltipIconButton'
import { HelpersContext } from '../../contexts'

const TodoList = ({}) => {
  const { showAlert } = useContext(HelpersContext)
  const [todoList, setTodoList] = useState([])
  const [isLoading, setLoading] = useState(false)
  const [showCompleted, setShowCompleted] = useState(false)
  const [needsUpdate, setNeedsUpdate] = useState(true)
  const [creating, setCreating] = useState(false)

  useEffect(() => {
    const getTodos = async () => {
      try {
        const endpoint = showCompleted ? 'todos/completed' : 'todos/active'
        const response = await API.get('instrument-inventory', endpoint)
        setTodoList(response)
      } catch (e) {
        console.error(e)
      }
    }
    if (needsUpdate) {
      getTodos()
      setNeedsUpdate(false)
    }
  }, [needsUpdate])

  const beforeAction = () => {
    setLoading(true)
  }

  const afterAction = () => {
    setNeedsUpdate(true)
    setLoading(false)
  }

  const createTodo = async data => {
    if (!data.content) {
      showAlert('Please enter some content')
      return
    }

    setLoading(true)
    try {
      await API.post('instrument-inventory', 'todos', {
        body: data,
      })
      showAlert('Todo Created')
    } catch (err) {
      showAlert(`Error: ${err}`)
    }

    setLoading(false)
    setNeedsUpdate(true)
    setCreating(false)
  }

  return (
    <React.Fragment>
      <LoadingHeader isLoading={isLoading} title="Todo List" />
      <br />
      <Table component="table" size="small">
        <TableHead component="thead">
          <TableRow component="tr">
            <TableCell component="th">Status</TableCell>
            <TableCell component="th">Task</TableCell>
            <TableCell component="th">Instrument</TableCell>
            <TableCell component="th">Actions</TableCell>
          </TableRow>
        </TableHead>
        <TableBody component="tbody">
          {creating ? (
            <TodoForm
              onSubmit={createTodo}
              icon={<AddIcon style={{ marginLeft: '3px' }} />}
              onCancel={() => setCreating(false)}
            />
          ) : (
            <TableRow component="tr" hover onClick={() => setCreating(true)}>
              <TableCell component="td">
                <AddIcon style={{ marginLeft: '3px' }} />
              </TableCell>
              <TableCell component="td">Create Todo</TableCell>
              <TableCell />
              <TableCell />
            </TableRow>
          )}
          {todoList.map(item => (
            <TodoItem
              item={item}
              key={item.todoId}
              beforeAction={beforeAction}
              afterAction={afterAction}
              showAlert={showAlert}
            />
          ))}
          <TableRow
            component="tr"
            hover
            onClick={() => {
              setShowCompleted(c => !c)
              setNeedsUpdate(true)
            }}
            data-testid="toggle-completed"
          >
            <TableCell component="td">
              <ArrowForwardIcon />
            </TableCell>
            <TableCell component="td">
              Show {showCompleted ? 'Incomplete' : 'Completed'}
            </TableCell>
            <TableCell />
            <TableCell />
          </TableRow>
        </TableBody>
      </Table>
    </React.Fragment>
  )
}

TodoList.propTypes = {}

export default TodoList

const TodoItem = ({ item, beforeAction, afterAction, showAlert }) => {
  const [editing, setEditing] = useState(false)
  const deleteTodo = async () => {
    beforeAction()
    try {
      await API.del('instrument-inventory', `todos/${item.todoId}`)
    } catch (err) {
      showAlert(`Error: ${err}`)
    }
    afterAction()
  }

  const toggleCompleted = async () => {
    beforeAction()
    try {
      const end = item.completed ? 'unmark-completed' : 'mark-completed'
      await API.post('instrument-inventory', `todos/${item.todoId}/${end}`)
    } catch (err) {
      showAlert(`Error: ${err}`)
    }
    afterAction()
  }

  const updateTodo = async data => {
    beforeAction()

    if (!data.content) {
      showAlert('Please enter some content')
      return
    }
    try {
      await API.put('instrument-inventory', `todos/${item.todoId}`, { body: data })
    } catch (err) {
      showAlert(`Error: ${err}`)
    }

    setEditing(false)
    afterAction()
  }

  return (
    <React.Fragment>
      {editing ? (
        <TodoForm
          onSubmit={updateTodo}
          icon={<EditIcon />}
          onCancel={() => setEditing(false)}
          initial={item}
        />
      ) : (
        <TableRow component="tr">
          <TableCell component="td">
            {item.completed ? (
              <TooltipIconButton
                title="Mark Incomplete"
                size="small"
                onClick={toggleCompleted}
              >
                <CheckCircle />
              </TooltipIconButton>
            ) : (
              <TooltipIconButton
                title="Mark Completed"
                size="small"
                onClick={toggleCompleted}
              >
                <CheckCircleOutline />
              </TooltipIconButton>
            )}
          </TableCell>
          <TableCell component="td">{item.content}</TableCell>
          <TableCell component="td">{item.relevantInstrument}</TableCell>
          <TableCell component="td">
            <TooltipIconButton
              title="Edit"
              size="small"
              onClick={() => setEditing(true)}
            >
              <EditIcon />
            </TooltipIconButton>
            <TooltipIconButton title="Delete" size="small" onClick={deleteTodo}>
              <DeleteIcon />
            </TooltipIconButton>
          </TableCell>
        </TableRow>
      )}
    </React.Fragment>
  )
}

TodoItem.propTypes = {
  item: PropTypes.object.isRequired,
  beforeAction: PropTypes.func,
  afterAction: PropTypes.func,
  showAlert: PropTypes.func.isRequired,
}

const TodoForm = ({
  initial = { content: '', relevantInstrument: '' },
  onCancel,
  onSubmit,
  icon,
}) => {
  const [content, setContent] = useState(initial.content)
  const [relevantInstrument, setRelevantInstrument] = useState(
    initial.relevantInstrument
  )

  return (
    <TableRow component="tr">
      <TableCell component="td">
        {icon}
        {/* <AddIcon style={{ marginLeft: '3px' }} /> */}
      </TableCell>
      <TableCell component="td">
        <TextField
          style={{ verticalAlign: 'center' }}
          label="Task"
          value={content}
          name="content"
          onChange={event => setContent(event.target.value)}
          required
          // error={error}
        />
      </TableCell>
      <TableCell component="td">
        <TextField
          label="Instrument"
          value={relevantInstrument}
          onChange={event => setRelevantInstrument(event.target.value)}
        />
      </TableCell>
      <TableCell component="td">
        <Button
          onClick={() => onSubmit({ content, relevantInstrument })}
          size="small"
          disabled={!content}
        >
          Submit
        </Button>
        <Button onClick={onCancel} size="small">
          Cancel
        </Button>
      </TableCell>
    </TableRow>
  )
}

TodoForm.propTypes = {
  initial: PropTypes.object,
  onCancel: PropTypes.func.isRequired,
  icon: PropTypes.element.isRequired,
  onSubmit: PropTypes.func.isRequired,
}
