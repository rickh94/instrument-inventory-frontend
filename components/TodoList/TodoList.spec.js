import React from 'react'
import TodoList from './TodoList'
import { render } from '@testing-library/react'
import { API } from 'aws-amplify'

describe('TodoList', () => {
  it('matches snapshot', () => {
    API.post = jest.fn().mockImplementation(() => Promise.resolve([{ id: 1 }]))
    const { container } = render(<TodoList showAlert={jest.fn()} />)
    expect(container).toMatchSnapshot()
  })
})
