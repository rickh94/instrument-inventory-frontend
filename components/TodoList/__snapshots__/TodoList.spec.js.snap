// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList matches snapshot 1`] = `
<div>
  <div
    class="makeStyles-heading-1"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 MuiTypography-colorInherit"
    >
      Todo List
    </h5>
  </div>
  <br />
  <table
    class="MuiTable-root"
  >
    <thead
      class="MuiTableHead-root"
    >
      <tr
        class="MuiTableRow-root MuiTableRow-head"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
          scope="col"
        >
          Status
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
          scope="col"
        >
          Task
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
          scope="col"
        >
          Instrument
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
          scope="col"
        >
          Actions
        </th>
      </tr>
    </thead>
    <tbody
      class="MuiTableBody-root"
    >
      <tr
        class="MuiTableRow-root MuiTableRow-hover"
      >
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            style="margin-left: 3px;"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
        >
          Create Todo
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
        />
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
        />
      </tr>
      <tr
        class="MuiTableRow-root MuiTableRow-hover"
      >
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
            />
          </svg>
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
        >
          Show 
          Completed
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
        />
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
        />
      </tr>
    </tbody>
  </table>
</div>
`;
